graveler/committed/sstable/cache.go

```diff
@@ -1,14 +1,103 @@
 package sstable
 
 import (
+	"fmt"
+
 	"github.com/cockroachdb/pebble/sstable"
+	lru "github.com/treeverse/lakefs/cache"
 	"github.com/treeverse/lakefs/graveler/committed"
+	"github.com/treeverse/lakefs/pyramid"
 )
 
 //go:generate mockgen -source=cache.go -destination=mock/cache.go -package=mock
 
+type Derefer lru.Derefer
+
+type opener func(namespace string, id committed.ID) (Item, error)
+
 type cache interface {
-	// derefer is the type of a function that returns an object to the cache, possibly causing
-	// its eviction.
-	GetOrOpen(namespace string, id committed.ID) (reader *sstable.Reader, derefer func() error, err error)
+	GetOrOpen(namespace string, id committed.ID) (*sstable.Reader, Derefer, error)
+}
+
+type lruCache struct {
+	c    lru.CacheWithDisposal
+	open opener
+}
+
+// item holds an SSTable inside a cache.  It exists (only) to allow tests to replace its
+// closer with a testable fake.
+type Item interface {
+	GetSSTable() *sstable.Reader
+	Close() error
+}
+
+type item struct {
+	r *sstable.Reader
+}
+
+func (i *item) GetSSTable() *sstable.Reader {
+	return i.r
+}
+
+func (i *item) Close() error {
+	return i.r.Close()
+}
+
+func NewCache(p lru.ParamsWithDisposal, fs pyramid.FS, readerOptions sstable.ReaderOptions) cache {
+	return NewCacheWithOpener(p, func(namespace string, id committed.ID) (Item, error) {
+		file, err := fs.Open(namespace, string(id))
+		if err != nil {
+			return nil, fmt.Errorf("fetch %s from next tier: %w", id, err)
+		}
+
+		reader, err := sstable.NewReader(file, readerOptions)
+		if err != nil {
+			return nil, fmt.Errorf("open SSTable %s: %w", id, err)
+		}
+		return &item{reader}, nil
+	})
+}
+
+func NewCacheWithOpener(p lru.ParamsWithDisposal, open opener) cache {
```

Comment:
If it's just for a test hook, maybe make it internal?

Reply:
We split up packages to separate test and non-test.  @nopcoder says to test at the caller level.

Also this avoids import cycles with mocks.  Otherwise, package mock depends on package x-for-code, and package x-for-test depends on package mock.  So if x-for-test is the same as x-for-code... we get an import cycle.